curvefs/src/metaserver/space/inode_volume_space_deallocate.cpp

```diff
@@ -0,0 +1,361 @@
+
+    for (const auto &inodeId : inodelist) {
+        VolumeExtentSliceList sliceList;
+        MetaStatusCode st = calOpt_.inodeStorage->GetAllVolumeExtent(
+            fsId_, inodeId, &sliceList);
+        if (st != MetaStatusCode::OK) {
+            LOG(ERROR) << "InodeVolumeSpaceDeallocate get inode extent "
+                          "failed, fsId="
+                       << fsId_ << "partitionId=" << partitionId_
+                       << ", inodeId=" << inodeId;
+            return false;
+        }
+        VLOG(6) << "InodeVolumeSpaceDeallocate will update inode:" << inodeId
+                << ", blockGroupOffset:" << blockGroupOffset
+                << ", sliceList:" << sliceList.DebugString();
+
+        // 1
+        UpdateDeallocateInodeExtentSlice(blockGroupOffset, inodeId, decrease,
+                                         &sliceList, &deallocatableVolumeSpace);
+        auto closure = new UpdateVolumeExtentClosure(nullptr, true);
+        executeOpt_.metaClient->AsyncUpdateVolumeExtent(fsId_, inodeId,
+                                                        sliceList, closure);
+        if (CURVEFS_ERROR::OK != closure->Wait()) {
+            LOG(ERROR) << "InodeVolumeSpaceDeallocate update inode:" << inodeId
+                       << ", blockGroupOffset:" << blockGroupOffset
+                       << ", to sliceList:" << sliceList.DebugString()
+                       << " failed";
+            return false;
+        }
+        VLOG(6) << "InodeVolumeSpaceDeallocate update inode:" << inodeId
+                << ", blockGroupOffset:" << blockGroupOffset
+                << ", to sliceList:" << sliceList.DebugString();
+    }
+
+    // 2
+    bool ok = executeOpt_.volumeSpaceManager->DeallocVolumeSpace(
+        fsId_, blockGroupOffset, deallocatableVolumeSpace);
+    if (!ok) {
+        // TODO(ilixiaocui): this part of the non-recyclable space should be
+        // included in the metric statistics
+        LOG(ERROR) << "InodeVolumeSpaceDeallocate deallocate volume space "
+                      "failed, fsId="
+                   << fsId_ << ", partitionId=" << partitionId_
+                   << ", blockGroupOffset=" << blockGroupOffset << " fail";
+    }
+
+    return ok;
+}
+
+void InodeVolumeSpaceDeallocate::UpdateDeallocateInodeExtentSlice(
+    uint64_t blockGroupOffset, uint64_t inodeId, uint64_t *decrease,
+    VolumeExtentSliceList *sliceList,
+    std::vector<Extent> *deallocatableVolumeSpace) {
+    uint64_t boundary =
+        (blockGroupOffset / blockGroupSize_ + 1) * blockGroupSize_;
+
+    for (auto &slice : *sliceList->mutable_slices()) {
+        auto extents = slice.mutable_extents();
+        extents->erase(
+            std::remove_if(
+                extents->begin(), extents->end(),
+                [&](const VolumeExtent &extent) {
+                    bool res = (extent.volumeoffset() >= blockGroupOffset) &&
+                               (extent.volumeoffset() < boundary);
```

Comment:
because of the existence of 4k headers, the length of the extent will not cross the block group, so there is no need to add length to judge.

Reply:
> because of the existence of 4k headers, the length of the extent will not cross the block group, so there is no need to add length to judge.

consider add a comment to explain this implicit condition